include/SH3/system/input.hpp

```diff
@@ -0,0 +1,533 @@
+/** @file
+ *  Defines the @ref sh3::system::input_system.
+ */
+#ifndef SH3_SYSTEM_INPUT_HPP_INCLUDED
+#define SH3_SYSTEM_INPUT_HPP_INCLUDED
+
+#include <bitset>
+#include <chrono>
+#include <cstdint>
+#include <ratio>
+#include <string>
+
+#include <boost/container/flat_map.hpp>
+#include <boost/operators.hpp>
+#include <boost/range/adaptor/map.hpp>
+#include <boost/range/adaptor/transformed.hpp>
+
+#include <SDL2/SDL_keyboard.h>
+#include <SDL2/SDL_mouse.h>
+#include <SDL2/SDL_timer.h>
+
+#include "input_config.hpp"
+
+struct vector2
+{
+    double x, y;
+};
+
+namespace sh3 { namespace system {
```

Comment:
I think it makes sense to use namespaces as we use directories.
And yes, many things we have in the SH3-directory, which is everything, are actually more generally applicable.

Moving these things about and changing to appropriate namespaces is probably a good project for someone wanting to get familiar with the source. I might have someone interested in contributing; I'll know in a week. I'll message you how it goes and if it doesn't work out either of us can do it instead.